packages/flutter_tools/lib/src/commands/create_utils.dart

```diff
@@ -0,0 +1,533 @@
+import '../convert.dart';
+import '../dart/pub.dart';
+import '../flutter_project_metadata.dart';
+import '../globals.dart' as globals;
+import '../project.dart';
+import '../runner/flutter_command.dart';
+import '../template.dart';
+
+/// Mixin contains methods that are shared with `flutter create` commands.
+mixin CreateCommandMixin on FlutterCommand {
+  /// Throw with exit code 2 if the output directory is invalid.
+  void validateOutoutDirectoryArg() {
+    if (argResults.rest.isEmpty) {
+      throwToolExit('No option specified for the output directory.\n$usage',
+          exitCode: 2);
+    }
+
+    if (argResults.rest.length > 1) {
+      String message = 'Multiple output directories specified.';
+      for (final String arg in argResults.rest) {
+        if (arg.startsWith('-')) {
+          message += '\nTry moving $arg to be immediately following $name';
+          break;
+        }
+      }
+      throwToolExit(message, exitCode: 2);
+    }
+  }
+
+  /// Gets the flutter root directory.
+  ///
+  /// Throw with exit code 2 if the flutter sdk installed is invalid.
+  String getFlutterRoot() {
+    if (Cache.flutterRoot == null) {
+      throwToolExit(
+          'Neither the --flutter-root command line flag nor the FLUTTER_ROOT environment '
+          'variable was specified. Unable to find package:flutter.',
+          exitCode: 2);
+    }
+    final String flutterRoot = globals.fs.path.absolute(Cache.flutterRoot);
+
+    final String flutterPackagesDirectory =
+        globals.fs.path.join(flutterRoot, 'packages');
+    final String flutterPackagePath =
+        globals.fs.path.join(flutterPackagesDirectory, 'flutter');
+    if (!globals.fs
+        .isFileSync(globals.fs.path.join(flutterPackagePath, 'pubspec.yaml'))) {
+      throwToolExit('Unable to find package:flutter in $flutterPackagePath',
+          exitCode: 2);
+    }
+
+    final String flutterDriverPackagePath =
+        globals.fs.path.join(flutterRoot, 'packages', 'flutter_driver');
+    if (!globals.fs.isFileSync(
+        globals.fs.path.join(flutterDriverPackagePath, 'pubspec.yaml'))) {
+      throwToolExit(
+          'Unable to find package:flutter_driver in $flutterDriverPackagePath',
+          exitCode: 2);
+    }
+    return flutterRoot;
+  }
+
+  /// Determines the project type in an existing flutter project.
+  ///
+  /// If it has a .metadata file with the project_type in it, use that.
+  /// If it has an android dir and an android/app dir, it's a legacy app
+  /// If it has an ios dir and an ios/Flutter dir, it's a legacy app
+  /// Otherwise, we don't presume to know what type of project it could be, since
+  /// many of the files could be missing, and we can't really tell definitively.
+  ///
+  /// Throws assertion if projectDir does not exist or empty.
+  /// Returns null if no project type can be determined.
+  FlutterProjectType determineTemplateType(Directory projectDir) {
+    assert(projectDir.existsSync() && projectDir.listSync().isNotEmpty);
```

Comment:
asserts only run during tests, do these need to be error checks?

Reply:
This method only makes sense when `projectDir` is not empty. So I think it makes sense to have an assertion here so bugs can be easier determined in the future. (And yes, it would only help debugging under tests)